### PR TITLE
Adds parameters to backup-auth for region support and non AWS endpoint support

### PR DESCRIPTION
**Pull request by krautnerds**
Hi everyone!

We recently started using [minio](https://github.com/minio/minio) as S3-compatbile microservice and want to use it now for our mariadb backups.

To get there this is the second pull request of two. The first one was already published and is a prerequisite for this pr:
https://github.com/dokku/s3gof3r-docker/pull/1

With this pr we can now use three new optional parameters for the backup-auth command:

1. AWS Region (new parameter in third place)
2. AWS Signature Version (new parameter in fourth place)
3. Different endpoints instead of only the default AWS S3 endpoint (new parameter in fifth place)

Special thanks to mbegenau for helping me out while coding these additions.

### CODE REVIEW

File: common-functions [hunk 1]
```diff
@@ -116,23 +115,51 @@ service_backup() {
...
+
+  if [ -f "$BACKUP_CONFIG_ROOT/AWS_SIGNATURE_VERSION" ]; then
+    BACKUP_PARAMETERS="$BACKUP_PARAMETERS -e AWS_SIGNATURE_VERSION=$(cat "$BACKUP_CONFIG_ROOT/AWS_SIGNATURE_VERSION")"
+  fi
+
+  if [ -f "$BACKUP_CONFIG_ROOT/ENDPOINT_URL" ]; then
+    BACKUP_PARAMETERS="$BACKUP_PARAMETERS -e ENDPOINT_URL=$(cat "$BACKUP_CONFIG_ROOT/ENDPOINT_URL")"
+  fi
+
+  docker run "$BACKUP_PARAMETERS" dokkupaas/s3backup:0.6.0
```
Comment (josegonzalez) [inline common-functions]: Are you sure you want quotes here?

File: common-functions [hunk 2]
```diff
@@ -116,23 +115,51 @@ service_backup() {
...
+  declare SERVICE="$1" AWS_ACCESS_KEY_ID="$2" AWS_SECRET_ACCESS_KEY="$3" AWS_DEFAULT_REGION="$4"
+  declare AWS_SIGNATURE_VERSION="$5" ENDPOINT_URL="$6"
   local SERVICE_ROOT="${PLUGIN_DATA_ROOT}/${SERVICE}"
   local SERVICE_BACKUP_ROOT="${SERVICE_ROOT}/backup/"
 
   mkdir -p "$SERVICE_BACKUP_ROOT"
   echo "$AWS_ACCESS_KEY_ID" > "${SERVICE_BACKUP_ROOT}/AWS_ACCESS_KEY_ID"
   echo "$AWS_SECRET_ACCESS_KEY" > "${SERVICE_BACKUP_ROOT}/AWS_SECRET_ACCESS_KEY"
+
+  if [ "x$AWS_DEFAULT_REGION" != "x" ]; then
```
Comment (josegonzalez) [inline common-functions]: This is not how we check for empty variables in bash. The following is preferred:

```shell
if [[ -n "$VAR" ]]; then
  echo "variable is non-zero length"
fi
```

File: common-functions [hunk 3]
```diff
@@ -116,23 +115,51 @@ service_backup() {
...
+  fi
+
+  docker run "$BACKUP_PARAMETERS" dokkupaas/s3backup:0.6.0
 }
 
 service_backup_auth() {
   declare desc="Sets up authentication"
-  declare SERVICE="$1" AWS_ACCESS_KEY_ID="$2" AWS_SECRET_ACCESS_KEY="$3"
+  declare SERVICE="$1" AWS_ACCESS_KEY_ID="$2" AWS_SECRET_ACCESS_KEY="$3" AWS_DEFAULT_REGION="$4"
+  declare AWS_SIGNATURE_VERSION="$5" ENDPOINT_URL="$6"
```
Comment (josegonzalez) [inline common-functions]: Combine these lines.

File: common-functions [hunk 4]
```diff
@@ -30,8 +30,7 @@ get_random_ports() {
     local port=$RANDOM
     local quit=0
     while [ "$quit" -ne 1 ]; do
-      netstat -an | grep $port > /dev/null
-      if [ $? -gt 0 ]; then
+      if ! netstat -an | grep $port > /dev/null; then
```
Comment (josegonzalez) [inline common-functions]: Why this change?

File: common-functions (hunk 1 above)
Comment (krautnerds) [inline common-functions]: I'm not shure, but your lint tool wanted them. :)

see https://travis-ci.org/dokku/dokku-mariadb/jobs/264175402

File: common-functions (hunk 4 above)
Comment (krautnerds) [inline common-functions]: your lint tool wanted that change ;)

see again: https://travis-ci.org/dokku/dokku-mariadb/jobs/264175402

File: common-functions (hunk 1 above)
Comment (josegonzalez) [inline common-functions]: You might need to disable that rule for this line.

Comment (krautnerds) [inline common-functions]: how do i do that?

Comment (josegonzalez) [inline common-functions]:
```shell
# shellcheck disable=SC2086
code_with_error_SC2086
```

File: common-functions (hunk 4 above)
Comment (krautnerds) [inline common-functions]: should i disable linting for this line too?

Comment (krautnerds) [inline common-functions]: Fixed.

File: common-functions (hunk 1 above)
Comment (krautnerds) [inline common-functions]: Fixed.

File: common-functions (hunk 3 above)
Comment (krautnerds) [inline common-functions]: Fixed.

File: common-functions (hunk 2 above)
Comment (krautnerds) [inline common-functions]: Fixed.

**Comment by krautnerds** [comment]
@josegonzalez the lint tool is also happy now ;)

File: common-functions [hunk 5]
```diff
@@ -116,23 +117,51 @@ service_backup() {
...
-      -v "${TMPDIR}:/backup" dokkupaas/s3backup:0.6.0
+
+  # Build parameter list for s3backup tool
+  BACKUP_PARAMETERS="-e AWS_ACCESS_KEY_ID=$(cat "$AWS_ACCESS_KEY_ID_FILE") \
+                     -e AWS_SECRET_ACCESS_KEY=$(cat "$AWS_SECRET_ACCESS_KEY_FILE") \
+                     -e BUCKET_NAME=$BUCKET_NAME \
+                     -e BACKUP_NAME=${PLUGIN_COMMAND_PREFIX}-${SERVICE} \
+                     -v ${TMPDIR}:/backup"
+
+  if [ -f "$BACKUP_CONFIG_ROOT/AWS_DEFAULT_REGION" ]; then
```
Comment (josegonzalez) [inline common-functions]: Two brackets, not one.

Comment (krautnerds) [inline common-functions]: Fixed. By the way: This convention should be added to your lint config :)

**Comment by krautnerds** [comment]
@josegonzalez I also updated to the newest version of dokkupaas/s3backup - thank you for accepting the first merge request :)

**Comment by krautnerds** [comment]
@josegonzalez is there anything more to do to get this into the next plugin version? Would love to use this in production!

I will also apply this patch set to all other plugins with AWS backup functionality if you like :)

**Comment by josegonzalez** [comment]
Yeah I basically have to go through and apply this + some other changes everywhere. Should be soon - I will have some time this weekend.

**Comment by krautnerds** [comment]
@josegonzalez Awesome! If there is anything I can help you to accelerate the publishing of the new version.
